org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/CompletionProposalRequestor.java

```diff
@@ -266,6 +266,11 @@ public CompletionItem toCompletionItem(CompletionProposal proposal, int index) {
 			}
 		}
 		proposalProvider.updateReplacement(proposal, $, '\0');
+		// Make sure `filterText` matches `textEdit`
+		// See https://github.com/eclipse/eclipse.jdt.ls/issues/1348
+		if (proposal.getKind() == CompletionProposal.TYPE_REF && $.getTextEdit() != null && $.getTextEdit().getRange() != null && $.getTextEdit().getNewText() != null) {
```

Comment:
Here to minimize the risk and avoit unexpected behaviors, I only apply it to `TYPE_REF` kind. In fact, if I remove `proposal.getKind() == CompletionProposal.TYPE_REF` condition, and update the `filterText` for all items, current test cases are also passed.